Inputting/InputBufferQueue.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Inputting
+{
+    /// <summary>
+    /// The input buffer queue is responsible for putting inputs in reference to time,
+    /// meaning when and for how long each input should be executed.
+    /// It allows for inputs to be queued and dequeued,
+    /// but dequeuing an input also returns a duration for that input.
+    /// </summary>
+    public class InputBufferQueue<T>
+    {
+        private readonly Queue<T> _queue = new Queue<T>();
+
+        private readonly float _bufferLengthSeconds;
+        private readonly float _speedupRate;
+        private readonly float _slowdownRate;
+        private readonly float _minInputDuration;
+        private readonly float _maxInputDuration;
+
+        private float _prevInputDuration;
+
+        private Queue<TaskCompletionSource<(T, float)>> _awaitedDequeueings =
+            new Queue<TaskCompletionSource<(T, float)>>();
+
+        /// <summary>
+        /// Create a new buffer for the given settings.
+        /// </summary>
+        /// <param name="bufferLengthSeconds">Duration in seconds the buffer queue aims to buffer inputs for.
+        /// E.g. if it is set to 3 seconds and inputs get enqueued and dequeued at a constant rate of 5 per second,
+        /// the queue will stabilize at 15 items.</param>
+        /// <param name="speedupRate">The rate at which input speed can increase,
+        /// as a percentage from 0 to 1 with 1 being instantaneous.</param>
+        /// <param name="slowdownRate">The rate at which input speed can decrease,
+        /// as a percentage from 0 to 1 with 1 being instantaneous.</param>
+        /// <param name="minInputDuration">The minimum duration a single input can have, in seconds.</param>
```

Comment:
because that's the SI unit. I could have it take a TimeSpan and add an extension method "FromFrames" to TimeSpan. But I thought assuming everything we ever run will be a locked 60fps game was a fragile assumption

Reply:
For most games, the input puller times to the frame. That should probably be our base unit.

Reply:
what's wrong with rounding the results*60 afterwards? I don't think this class needs to deal with that

Reply:
I guess I prefer integer math to floating point math

Reply:
Can't you just take milliseconds then?